blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc

```diff
@@ -0,0 +1,701 @@
+before acknowledging the retrieval of a specific event with the source database or the messaging broker, respectively.
+In that case, after a restart of Debezium or the consuming service,
+a few events may be processed a second time.
+Propagating the event UUID as a Kafka message header allows for an efficient detection and exclusion duplicate in the consumer.
+
+If a message is received for the first time, the message value is parsed and the business method of the `ShippingService` method corresponding to the specific event type is invoked with the event payload.
+Finally, the message is marked as processed with the message log.
+
+This `MessageLog` simply keeps track of all consumed events in a table within the service's local database:
+
+[source,java]
+----
+@ApplicationScoped
+public class MessageLog {
+
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    @Transactional(value=TxType.MANDATORY)
+    public void processed(UUID eventId) {
+        entityManager.persist(new ConsumedMessage(eventId, Instant.now()));
+    }
+
+    @Transactional(value=TxType.MANDATORY)
+    public boolean alreadyProcessed(UUID eventId) {
+        return entityManager.find(ConsumedMessage.class, eventId) != null;
+    }
+}
+----
+
+That way, should the transaction be rolled back for some reason, also the original message will not be marked as processed and an exception would bubble up to the Kafka event consumer loop.
+This allows for re-trying to process the message later on.
+
+Note that a more complete implementation should take care of re-trying given messages only for a certain number of times,
+before re-routing any unprocessable messages to a dead-letter queue or similar.
+Also there should be some house-keeping on the message log table;
+periodically, all events older than the consumer's current offset committed with the broker may be deleted,
+as it's ensured that such messages won't be propagated to the consumer another time.
+
+== Summary
+
+The outbox pattern is a great way for propagating data amongst different microservices.
+
+By only modifying a single resource - the source service's own database -
+it avoids any potential inconsistencies of altering multiple resources at the same time which don't share one common transactional context
+(the database and Apache Kafka).
+By writing to the database first, the source service has instant "read your own writes" semantics,
+which is important for a consistent user experience, allowing query methods invoked following to a write to instantly reflect any data changes.
+
+At the same time, the pattern enables asynchronous event propagation to other microservices.
+Apache Kafka acts as a highly scalable and reliable backbone for the messaging amongst the services.
+Given the right topic retention settings, new consumers may come up long after an event has been originally produced,
+and build up their own local state based on the event history.
```

Comment:
This is again subject to the same "there are options here" qualification.

Reply:
Should be addressed now.